language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{counters::*, create_access_path, data_cache::StateViewCache};
+use anyhow::Result;
+use diem_state_view::StateView;
+use diem_types::{
+    account_address::AccountAddress,
+    account_config::{self, CurrencyInfoResource, RoleId},
+    transaction::{
+        GovernanceRole, SignatureCheckedTransaction, SignedTransaction, VMValidatorResult,
+    },
+    vm_status::{StatusCode, VMStatus},
+};
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    move_resource::MoveStructType,
+};
+use move_vm_runtime::{data_cache::MoveStorage, session::Session};
+
+use crate::logging::AdapterLogSchema;
+
+/// This trait describes the VM adapter's interface.
+pub trait VMAdapter {
+    /// Creates a new Session backed by the given storage.
+    /// TODO: this probably doesn't belong in this trait. We should be able to remove
```

Comment:
I had a chat with @vgao1996 and we agreed that there are issues with the session ownership model in the vm. He plans to get rid of session, and have the clients (of the vm) own the caches. For now, we can probably leave session here as is.

Reply:
Yeah I tried to propose some cleaner APIs to Emma, but then realized those would require an additional cache flush with the way the code cache is currently set up. It will be hard to implement those without fixing the ownership model first. Will look into it once I get back from vacation.